File: 39-combination-sum/combination-sum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        result = []

        def backtrack(current, remaining, start):
            if remaining < 0:
                return
            elif remaining == 0:
                result.append(current[:]) 
            else:
                for i in range(start, len(candidates)):
                    current.append(candidates[i])
                    backtrack(current, remaining - candidates[i], i)
                    current.pop() 

        candidates.sort()
        backtrack([], target, 0)

        return result

File: 39-combination-sum/combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result = new ArrayList<>();

        Arrays.sort(candidates);

        backtrack(result, new ArrayList<>(), candidates, target, 0);
        return result;
    }

    private void backtrack(List<List<Integer>> result, List<Integer> current, int[] candidates, int remaining, int start) {
        if (remaining < 0) {
            return;
        } else if (remaining == 0) {
            result.add(new ArrayList<>(current));
        } else {
            for (int i = start; i < candidates.length; i++) {
                current.add(candidates[i]);
                backtrack(result, current, candidates, remaining - candidates[i], i);
                current.remove(current.size() - 1); // 백트래킹: 마지막 요소 제거
            }
        }
    }

}